// node-cron/dist/node-cron.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var inlineScheduledTask = require('./tasks/inline-scheduled-task.cjs');
var taskRegistry = require('./task-registry.cjs');
var logger = require('./logger.cjs');
var patternValidation = require('./pattern/validation/pattern-validation.cjs');
var backgroundScheduledTask = require('./tasks/background-scheduled-task/background-scheduled-task.cjs');
var path = require('path');
var url = require('url');

const moduleFilename = __filename;
const registry = new taskRegistry.TaskRegistry();
function schedule(expression, func, options) {
    const task = createTask(expression, func, options);
    const started = task.start();
    if (started && typeof started.catch === 'function') {
        started.catch((error) => {
            (options?.logger || logger.default).error(`Failed to start scheduled task: ${error?.message ?? error}`);
        });
    }
    return task;
}
function createTask(expression, func, options) {
    let task;
    if (func instanceof Function) {
        task = new inlineScheduledTask.InlineScheduledTask(expression, func, options);
    }
    else {
        const taskPath = solvePath(func);
        task = new backgroundScheduledTask.default(expression, taskPath, options);
    }
    registry.add(task);
    return task;
}
function solvePath(filePath) {
    if (path.isAbsolute(filePath))
        return url.pathToFileURL(filePath).href;
    if (filePath.startsWith('file://'))
        return filePath;
    const stackLines = new Error().stack?.split('\n');
    if (stackLines) {
        stackLines?.shift();
        const callerLine = stackLines?.find((line) => { return line.indexOf(moduleFilename) === -1; });
        const match = callerLine?.match(/(file:\/\/)?(((\/?)(\w:))?([/\\].+)):\d+:\d+/);
        if (match) {
            const dir = `${match[5] ?? ""}${path.dirname(match[6])}`;
            return url.pathToFileURL(path.resolve(dir, filePath)).href;
        }
    }
    throw new Error(`Could not locate task file ${filePath}`);
}
function validate(expression) {
    try {
        patternValidation.default(expression);
        return true;
    }
    catch (e) {
        return false;
    }
}
const getTasks = registry.all;
const getTask = registry.get;
const nodeCron = {
    schedule,
    createTask,
    validate,
    getTasks,
    getTask,
    setLogger: logger.setLogger,
};

exports.setLogger = logger.setLogger;
exports.createTask = createTask;
exports.default = nodeCron;
exports.getTask = getTask;
exports.getTasks = getTasks;
exports.nodeCron = nodeCron;
exports.schedule = schedule;
exports.solvePath = solvePath;
exports.validate = validate;
